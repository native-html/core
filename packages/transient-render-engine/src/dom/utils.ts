import { Text as TextType, Tag as TagType } from 'domelementtype';
import { Node, Text, Element } from 'domhandler';

export function isTextNode(node: Node | null | undefined): node is Text {
  return !!node && node.type === TextType;
}

export function isElementNode(node: Node | null | undefined): node is Element {
  return !!node && node.type === TagType;
}
