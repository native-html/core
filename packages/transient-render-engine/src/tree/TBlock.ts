import { TNode, TNodeInit } from './TNode';
import { SerializableElement } from '../dom/to-serializable';
import HTMLContentModel from '../model/HTMLContentModel';

export interface TBlockInit extends TNodeInit {}

export class TBlock extends TNode {
  public domNode?: SerializableElement;
  public readonly displayName: string = 'TBlock';
  constructor(init: TBlockInit) {
    super(init, 'block');
    if (init.domNode) {
      this.domNode = init.domNode;
    }
  }

  matchContentModel(contentModel: HTMLContentModel) {
    return (
      contentModel === HTMLContentModel.block ||
      contentModel === HTMLContentModel.mixed
    );
  }
}
