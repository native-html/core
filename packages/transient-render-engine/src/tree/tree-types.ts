import { Element, Text } from '../dom/dom-utils';
import HTMLContentModel from '../model/HTMLContentModel';
import HTMLElementModel from '../model/HTMLElementModel';
import { TStylesShape } from '../styles/TStyles';
import { TStylesMerger } from '../styles/TStylesMerger';

export interface TNodeInit {
  readonly elementModel: HTMLElementModel<string, HTMLContentModel> | null;
  readonly context: {
    readonly stylesMerger: TStylesMerger;
    readonly setMarkersForTNode?: SetMarkersForTNode;
    readonly removeLineBreaksAroundEastAsianDiscardSet: boolean;
  };
  readonly parent?: TNodeImpl | null;
  readonly domNode?: Element | null;
  readonly styles?: TStylesShape | null;
  readonly parentStyles?: TStylesShape | null;
  readonly nodeIndex?: number;
  readonly isUnregistered?: boolean;
}

/**
 * A minimal TNode shape accessible at instantiation.
 *
 * @public
 */
export interface TNodeDescriptor {
  tagName: string | null;
  classes: string[];
  id: string | null;
  attributes: Record<string, string>;
  markers: Markers;
}

/**
 * A function to set markers for a specific {@link TNode}.
 *
 * @param targetMarkers - The markers instance to alter.
 * @param parentMarkers - Th direct ascendant markers instance for this tnode.
 * @param tnode - The minimalistic {@link TNodeDescriptor} to which the
 * `targetMarkers` will be attached.
 *
 * @public
 */
export type SetMarkersForTNode = (
  targetMarkers: Markers,
  parentMarkers: Readonly<Markers>,
  tnode: TNodeDescriptor
) => void;

/**
 * Markers form an abstraction in which one node provides semantic information
 * to itself and all its descendants. For example, `ins` elements, which stand
 * for "insertion" of content in the context of an edit will provide the {
 * edits: 'ins' } marker to all its descendants.
 *
 * @remarks
 * Custom renderers can use markers to change their layout and convey their
 * semantic meaning. Markers can be derived from attributes, such as `lang` and
 * `dir` attributes, or tag names, such as `a`, `ins`, `del`...
 *
 * **TypeScript users**: You can add fields to the {@link Markers} interface via
 * {@link https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation | module augmentation}:
 *
 * ```ts
 * declare module '@native-html/transient-render-engine' {
 *   interface Markers {
 *     customField?: boolean;
 *   }
 * }
 * ```
 */
export interface Markers {
  /**
   * If this node is an `a` or has one as ancestor, this field will be set to
   * `true`.
   *
   * @defaultValue false
   */
  anchor: boolean;
  /**
   * If this node is a `del` or `ins` or has either as ancestor, this field
   * will be set accordingly. Otherwise, it will be set to 'none'.
   *
   * https://html.spec.whatwg.org/#edits
   *
   * @defaultValue 'none'
   */
  edits: 'ins' | 'del' | 'none';
  /**
   * The direction for this content. Follows `dir` attribute.
   *
   * https://html.spec.whatwg.org/#the-dir-attribute
   *
   * @defaultValue 'ltr'
   */
  direction: 'ltr' | 'rtl';
  /**
   * The language for this content. Follows `lang` attribute.
   *
   * https://html.spec.whatwg.org/#the-lang-and-xml:lang-attributes
   */
  lang: string;
  /**
   * * -1: this node is not an `ol` and has no `ol` parents
   * * 0: this node is an `ol` or has one `ol` parent
   * * 1: ...
   */
  olNestLevel: number;
  /**
   * * -1: this node is not an `ul` and has no `ul` parents
   * * 0: this node is an `ul` or has one `ul` parent
   * * 1: ...
   */
  ulNestLevel: number;
  /**
   * Create a new marker instance which extends this markers.
   */
  extend(this: Markers): Markers;
  /**
   * Return a string represenation of this marker, including inherited
   * properties from parent markers.
   */
  toString(): string;
}

/**
 * @public
 */
export interface TNodePrintOptions {
  /**
   * Include styles in snapshot.
   */
  withStyles: boolean;
  /**
   * Include node index in snapshot.
   */
  withNodeIndex: boolean;
}

/**
 * Processed styles which can be passed to `View` elements.
 *
 * @public
 */
export type NativeBlockStyles = TStylesShape['nativeBlockFlow'] &
  TStylesShape['nativeBlockRet'];

/**
 * Processed styles which can be passed to `Text` elements.
 *
 * @public
 */
export type NativeTextStyles = TStylesShape['nativeBlockFlow'] &
  TStylesShape['nativeBlockRet'] &
  TStylesShape['nativeTextFlow'] &
  TStylesShape['nativeTextRet'];

/**
 * Processed Web text styles.
 *
 * @public
 */
export type WebTextStyles = TStylesShape['webTextFlow'];

/**
 * Processed Web block styles.
 *
 * @public
 */
export type WebBlockStyles = TStylesShape['webBlockRet'];

/**
 * Shared properties for all {@link TNode} types.
 *
 * @public
 */
export interface TNodeShape<T extends TNodeType> {
  /**
   * Attributes for this tag.
   */
  readonly attributes: Record<string, string>;
  /**
   * The id for this node, extracted from the id attribute of the
   * underlying {@link Node}.
   */
  readonly id: string | null;
  /**
   * A list of classes for this node, extracted from the class attribute of the
   * underlying {@link Node}.
   */
  readonly classes: string[];
  /**
   * Non-anonymous nodes will hold a reference to a DOM node.
   */
  readonly domNode: Element | null;
  /**
   * The tag name for this node.
   *
   * @remarks Anonymous nodes generated during hoisting won't have a tag name.
   * Also, some TText nodes don't have a tagName.
   */
  readonly tagName: string | null;
  /**
   * The parent of this node before hoisting.
   *
   * @remarks
   * "Before hoisting" implies that this parent will not match "anonymous"
   * parents.
   */
  readonly parent: TDocument | TBlock | TPhrasing | null;
  /**
   * The position of this element relatively to its parents, before hoisting,
   * after collapsing.
   *
   * @remarks
   * "Before hoisting" implies that this index corresponds to the node position
   * in the DOM, after removal of empty elements as per whitespace collapsing
   * algorithm.
   */
  readonly nodeIndex: number;
  /**
   * Children of this node.
   */
  readonly children: ReadonlyArray<TNode>;

  /**
   * Used for debugging purposes.
   */
  readonly displayName: string;

  /**
   * The type of this tnode.
   */
  readonly type: T;

  /**
   * `true` when this tag is not a valid HTML tag **and** there is no custom
   * renderer for this tag.
   */
  readonly isUnregistered: boolean;

  /**
   * Markers form an abstraction in which one node provides semantic information
   * to itself and all its descendants. For example, `ins` elements, which stand
   * for "insertion" of content in the context of an edit will provide the {
   * edits: 'ins' } marker to all its descendants.
   *
   * @remarks This attribute must be considered immutable. Never try to
   * change it by hand, or you might update the markers of an anscestor! For
   * performance reasons, markers instances might be shared from parent to
   * children when they are equal.
   */
  readonly markers: Readonly<Markers>;

  /**
   * TStyles for this TNode. You should not use these unless required.
   * Use {@link TNodeShape.getNativeStyles} instead.
   */
  readonly styles: TStylesShape;

  /**
   * Create a JSX string representation of this node and its children.
   *
   * @remarks The snapshot is _just_ a representation. For example, it will
   * print `href` as an attribute of the TNode, while if you want to access
   * `href` programatically, you'll need to access it via
   * `tnode.attributes.href`.
   *
   * @param options - Customize what should be displayed.
   */
  snapshot(options?: Partial<TNodePrintOptions>): string;

  /**
   * Check if this {@link TNode} has the given `className`.
   *
   * @param className - The class to check.
   */
  hasClass(className: string): boolean;

  /**
   * Test if the given content model matches this TNode content model.
   *
   * @param contentModel - The content model to test against.
   */
  matchContentModel(contentModel: HTMLContentModel): boolean;

  /**
   * Get own native styles.
   */
  getNativeStyles(): T extends 'text' | 'phrasing'
    ? NativeTextStyles
    : NativeBlockStyles;

  /**
   * Get styles that cannot be handled by React Native.
   */
  getWebStyles(): WebTextStyles & WebBlockStyles;
}

/**
 * Metainformation extracted from `<head>` tag.
 *
 * @public
 */
export interface DocumentContext {
  charset: string;
  baseHref: string;
  baseTarget: '_blank' | '_self' | '_parent' | '_top';
  lang: string;
  dir: 'ltr' | 'rtl';
  title: string;
  meta: { name: string; value: string }[];
  links: Record<string, string>[];
}

/**
 * A transient render node (TNode) is a ready-to-render data structure.
 */
export type TNode = TBlock | TDocument | TEmpty | TPhrasing | TText;

/**
 * Transient render nodes to be rendered in React Native `Views`.
 */
export interface TBlock extends TNodeShape<'block'> {
  readonly tagName: string;
  readonly domNode: Element;
}

/**
 * The root of a transient render tree. It has a special
 * `context` field holding metadata about the page.
 */
export interface TDocument extends TNodeShape<'document'> {
  /**
   * An object containing special information for this document, such as lang,
   * dir, charset, baseHref...
   */
  readonly context: Readonly<DocumentContext>;
}

/**
 * Transient render nodes for a non-renderable elements, e.g.
 * for elements which have an {@link HTMLContentModel} set to **none**.
 */
export interface TEmpty extends TNodeShape<'empty'> {
  readonly domNode: Element;
}

/**
 * Transient render nodes to be rendered in React Native `Text`
 * components. TPhrasing nodes have {@link TText}, {@link TPhrasing} or
 * {@link TEmpty} children.
 */
export interface TPhrasing extends TNodeShape<'phrasing'> {}

/**
 * Transient render nodes to be rendered in React Native `Text`
 * components. They don't have children, and have a special `data` field
 * holding the text to be rendered.
 */
export interface TText extends TNodeShape<'text'> {
  readonly data: string;
  readonly textNode: Text;
}

/**
 * Distinct {@link TNode} types.
 */
export type TNodeType = 'block' | 'phrasing' | 'text' | 'empty' | 'document';

export interface TNodeMethods {
  bindChildren(
    this: TNodeImpl,
    children: ReadonlyArray<TNodeImpl>,
    shouldUpdateNodeIndexes?: boolean
  ): void;
  cloneInitParams<T extends TNodeInit = TNodeInit>(
    this: TNodeImpl,
    partial?: Partial<T>
  ): T;
  isCollapsibleLeft(this: TNodeImpl): boolean;
  isCollapsibleRight(this: TNodeImpl): boolean;
  isEmpty(this: TNodeImpl): boolean;
  trimLeft(this: TNodeImpl): void;
  trimRight(this: TNodeImpl): void;
  matchContentModel(contentModel: HTMLContentModel): boolean;
  collapse(this: TNodeImpl): void;
  hasClass(className: string): boolean;
  /**
   * Collpase this node children.
   *
   * @param params
   * @returns A list of indexes to splice
   */
  collapseChildren(this: TNodeImpl): void;
  spliceChildren(this: TNodeImpl, indexes: number[]): void;
  snapshot(options?: Partial<TNodePrintOptions>): string;
  toString(): string;
  initialize(init: TNodeInit): void;
  setMarkers: SetMarkersForTNode;
}

export interface TNodeImpl<T = TNodeInit>
  extends TNodeMethods,
    Omit<TNodeShape<TNodeType>, 'children' | 'parent'> {
  __nodeIndex: number | null;
  __trimmedLeft: boolean;
  __trimmedRight: boolean;
  readonly children: ReadonlyArray<TNodeImpl>;
  readonly init: T;
  readonly parent: TNodeImpl | null;
  readonly hasWhiteSpaceCollapsingEnabled: boolean;
  readonly parentStyles: TStylesShape | null;
  readonly elementModel: HTMLElementModel<string, HTMLContentModel> | null;
  readonly contentModel: HTMLContentModel | null;
}
