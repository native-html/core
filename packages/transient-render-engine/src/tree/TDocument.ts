import { TNodeInit } from './TNode';
import { TEmpty } from './TEmpty';
import { TBlock } from './TBlock';
import {
  isSerializableElement,
  isSerializableText
} from '../dom/to-serializable';
import HTMLElementModel from '../model/HTMLElementModel';
import defaultHTMLModelRecord from '../model/defaultHTMLModelRecord';

export interface DocumentContext {
  charset: string;
  baseHref: string;
  baseTarget: '_blank' | '_self' | '_parent' | '_top';
  lang: string;
  title: string;
  meta: { name: string; value: string }[];
  links: Record<string, string>[];
}

const defaultContextBase: DocumentContext = Object.freeze({
  baseHref: 'about:blank',
  baseTarget: '_self',
  charset: 'utf-8',
  title: '',
  lang: 'en',
  links: [],
  meta: []
});

function getDefaultDocumentContext(): DocumentContext {
  return Object.assign({}, defaultContextBase, { links: [], meta: [] });
}

function extractContextFromHead(head: TEmpty, lang?: string) {
  const context = getDefaultDocumentContext();
  if (lang) {
    context.lang = lang;
  }
  const domNode = head.domNode;
  const children = domNode.children;
  children.filter(isSerializableElement).forEach((child) => {
    if (child.tagName === 'meta') {
      if (child.attribs.name) {
        context.meta.push(child.attribs as any);
      } else if (child.attribs.charset) {
        context.charset = child.attribs.charset.toLowerCase();
      }
    } else if (child.tagName === 'link') {
      context.links.push(child.attribs);
    } else if (child.tagName === 'title') {
      for (const titleChild of child.children.filter(isSerializableText)) {
        context.title = titleChild.data.trim();
        break;
      }
    } else if (child.tagName === 'base') {
      context.baseHref = child.attribs.href || context.baseHref;
      context.baseTarget = (child.attribs.target as any) || context.baseTarget;
    }
  });
  return context;
}

const htmlModel = HTMLElementModel.fromNativeModel({
  tagName: 'html' as any,
  category: 'grouping'
});

export class TDocument extends TBlock {
  public readonly context: Readonly<DocumentContext>;
  public readonly displayName = 'TDocument';
  constructor({
    attributes,
    stylesMerger,
    parentStyles
  }: Omit<TNodeInit, 'contentModel' | 'elementModel'>) {
    super({
      tagName: 'html',
      attributes,
      contentModel: htmlModel.contentModel,
      elementModel: htmlModel,
      parentStyles: parentStyles,
      stylesMerger
    });
    // @ts-ignore
    this.type = 'document';
    this.context = defaultContextBase;
  }

  /**
   * Iterate over children and extract meta-information into context field.
   * Replace children with a single-element array containing the body.
   */
  parseChildren() {
    let head: TEmpty | undefined;
    let body: TBlock | undefined;
    for (const child of this.children) {
      if (head && body) {
        break;
      }
      if (child.tagName === 'head') {
        head = child as TEmpty;
      } else if (child.tagName === 'body') {
        body = child;
      }
    }
    this.bindChildren([
      body ||
        new TBlock({
          tagName: 'body',
          contentModel: defaultHTMLModelRecord.body.contentModel,
          elementModel: defaultHTMLModelRecord.body,
          stylesMerger: this.stylesMerger,
          parentStyles: this.styles
        })
    ]);
    //@ts-ignore
    this.context = Object.freeze(
      extractContextFromHead(
        head ||
          new TEmpty({
            tagName: 'head',
            isUnregistered: false,
            stylesMerger: this.stylesMerger,
            parentStyles: null,
            contentModel: defaultHTMLModelRecord.head.contentModel,
            elementModel: defaultHTMLModelRecord.head,
            domNode: {
              type: 'element',
              attribs: {},
              children: [],
              tagName: 'head'
            }
          }),
        this.attributes.lang
      )
    );
  }
}
