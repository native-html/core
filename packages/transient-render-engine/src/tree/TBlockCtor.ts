import HTMLContentModel from '../model/HTMLContentModel';
import TNodeCtor, { GenericTNodeCtor, Mutable } from './TNodeCtor';
import { TNodeImpl, TNodeInit } from './tree-types';

export interface TBlockImpl extends TNodeImpl {}

const TBlockCtor = function TBlock(this: Mutable<TBlockImpl>, init: TNodeInit) {
  this.initialize(init);
} as Function as GenericTNodeCtor<TNodeInit, TBlockImpl>;

//@ts-ignore
TBlockCtor.prototype = new TNodeCtor('block', 'TBlock');

TBlockCtor.prototype.matchContentModel = function matchContentModel(
  contentModel
) {
  return (
    contentModel === HTMLContentModel.block ||
    contentModel === HTMLContentModel.mixed
  );
};

TBlockCtor.prototype.collapseChildren = function collapseChildren() {
  let indexesToSplice: number[] = [];
  this.children.forEach((child, i) => {
    child.collapse();
    if (child.isEmpty()) {
      indexesToSplice.push(i);
    }
  });
  this.spliceChildren(indexesToSplice);
};

TBlockCtor.prototype.getNativeStyles = function getNativeStyles() {
  return {
    ...this.styles.nativeBlockFlow,
    ...this.styles.nativeBlockRet
  };
};

export default TBlockCtor;

export { TBlockCtor };
