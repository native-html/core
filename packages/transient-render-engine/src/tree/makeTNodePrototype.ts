import { TNodeImpl, TNodeInit, TNodeInvariants, TNodeType } from './tree-types';

export type TNodeCtor<Init = TNodeInit, Impl = TNodeImpl> = {
  new (init: Init): Impl;
  prototype: Impl;
};

export type Mutable<T> = {
  -readonly [P in keyof T]: T[P];
};

function updateNodeIndexes(node: Mutable<TNodeImpl>, i: number) {
  node.nodeIndex = i;
}

const emptyAttrs = {};

const prototype: Omit<TNodeImpl, keyof TNodeInvariants> & {
  // TODO remove
  className: string;
} = {
  children: Object.freeze([]) as any,
  init: Object.freeze({}) as any,
  __classes: null,
  __styles: null,
  __nodeIndex: null,
  __trimmedLeft: false,
  __trimmedRight: false,
  get attributes() {
    return this.domNode?.attribs || emptyAttrs;
  },

  get hasWhiteSpaceCollapsingEnabled() {
    return typeof this.styles.webTextFlow.whiteSpace === 'string'
      ? this.styles.webTextFlow.whiteSpace === 'normal'
      : true;
  },

  get contentModel() {
    return this.elementModel?.contentModel || null;
  },

  get parentStyles() {
    return this.init.parentStyles || this.parent?.styles || null;
  },

  get id() {
    return this.attributes.id || null;
  },

  get classes() {
    if (this.__classes === null) {
      this.__classes = this.attributes.class?.split(/\s+/) || [];
    }
    return this.__classes;
  },

  get className() {
    // TODO remove
    return this.attributes.class;
  },

  get domNode() {
    return this.init.domNode || null;
  },

  get elementModel() {
    return this.init.elementModel;
  },

  get stylesMerger() {
    return this.init.stylesMerger;
  },

  get styles() {
    if (this.__styles === null) {
      const self = (this as unknown) as TNodeImpl;
      this.__styles =
        self.init.styles ||
        self.init.stylesMerger.buildStyles(
          self.attributes.style,
          self.parentStyles || null,
          (this as unknown) as any
        );
    }
    return this.__styles;
  },

  get tagName() {
    return this.init.domNode?.name || null;
  },

  get parent() {
    return this.init.parent || null;
  },

  get nodeIndex() {
    if (this.__nodeIndex === null) {
      this.__nodeIndex = this.init.nodeIndex || 0;
    }
    return this.__nodeIndex;
  },

  set nodeIndex(nodeIndex: number) {
    this.__nodeIndex = nodeIndex;
  },

  bindChildren(children, shouldUpdateNodeIndexes = false) {
    //@ts-ignore
    this.children = children;
    if (shouldUpdateNodeIndexes) {
      children.forEach(updateNodeIndexes);
    }
  },

  cloneInitParams(partial) {
    return Object.assign({}, this.init, partial) as any;
  },

  isCollapsibleLeft() {
    if (this.children.length) {
      return (
        this.hasWhiteSpaceCollapsingEnabled &&
        this.children[0].isCollapsibleLeft()
      );
    }
    return false;
  },

  isCollapsibleRight() {
    if (this.children.length) {
      return (
        this.hasWhiteSpaceCollapsingEnabled &&
        this.children[this.children.length - 1].isCollapsibleRight()
      );
    }
    return false;
  },

  isWhitespace() {
    return false;
  },

  isEmpty() {
    return false;
  },

  trimLeft() {
    if (!this.__trimmedLeft && this.children.length) {
      const firstChild = this.children[0];
      firstChild.trimLeft();
      if (firstChild.isEmpty()) {
        //@ts-ignore
        this.children.splice(0, 1);
      }
      this.__trimmedLeft = true;
    }
  },

  trimRight() {
    if (!this.__trimmedRight && this.children.length) {
      const lastChild = this.children[this.children.length - 1];
      lastChild.trimRight();
      if (lastChild.isEmpty()) {
        //@ts-ignore
        this.children.splice(-1, 1);
      }
      this.__trimmedRight = true;
    }
  },

  matchContentModel() {
    return false;
  },

  spliceChildren(indexesToSplice) {
    if (indexesToSplice) {
      let offset = 0;
      for (const i of indexesToSplice) {
        //@ts-ignore
        this.children.splice(i - offset, 1);
        offset += 1;
      }
    }
  },

  collapse(params) {
    this.collapseChildren(params);
    this.bindChildren(this.children, true);
  },

  collapseChildren() {
    return;
  }
};

export function initialize<Impl extends TNodeImpl<any> = TNodeImpl>(
  self: Mutable<Impl>,
  init: Impl['init']
) {
  // self.init = init;
  Object.defineProperty(self, 'init', {
    get: () => init,
    enumerable: false
  });
}

const PCtor = (function PCtor(
  this: Mutable<TNodeImpl>,
  type: TNodeType,
  displayName: string
) {
  this.type = type;
  this.displayName = displayName;
} as unknown) as {
  new (type: TNodeType, displayName: string): TNodeImpl;
  prototype: TNodeImpl;
};

PCtor.prototype = (prototype as unknown) as TNodeImpl;

export default function makeTNodePrototype<Impl = TNodeImpl>(
  type: TNodeType,
  displayName: string,
  extraAccessors?: {
    [k in Exclude<keyof Impl, keyof TNodeImpl>]?: {
      get: () => any;
      set?: (val: any) => void;
    };
  }
): Impl {
  //@ts-ignore
  return Object.create(new PCtor(type, displayName), extraAccessors);
}
