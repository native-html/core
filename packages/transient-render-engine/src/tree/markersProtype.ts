import { Markers } from './tree-types';

const markersProtype: Markers = {
  anchor: false,
  edits: 'none',
  lang: 'en',
  olNestLevel: -1,
  ulNestLevel: -1,
  direction: 'ltr',
  extend(this: Markers) {
    return Object.create(this);
  },
  toString() {
    let print = 'Markers {\n';
    for (const key in this) {
      //@ts-ignore
      const val = this[key];
      if (typeof val !== 'function') {
        print += `  ${key}: ${JSON.stringify(val)}\n`;
      }
    }
    return print + '}';
  }
};

export default markersProtype;
