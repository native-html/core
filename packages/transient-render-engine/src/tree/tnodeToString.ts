import { TStyles } from '../styles/TStyles';
import { TNode, TNodePrintOptions } from './tree-types';

function printTableStyles(styles: TStyles) {
  const allStyles = {
    ...styles.nativeTextFlow,
    ...styles.nativeBlockFlow,
    ...styles.webTextFlow,
    ...styles.nativeTextRet,
    ...styles.nativeBlockRet
  };
  const entries = Object.entries(allStyles);
  return entries.reduce((prev, [name, val]) => {
    return `${prev}${prev ? ',' : ''} ${name}: ${JSON.stringify(val)}`;
  }, '');
}

function tnodePropertiesString(tnode: TNode, withStyles: boolean) {
  const tagPrint = tnode.tagName ? `tagName="${tnode.tagName}"` : 'anonymous';
  const unregisteredPrint = tnode.isUnregistered ? 'unregistered' : null;
  const indexPrint = `nodeIndex={${tnode.nodeIndex}}`;
  const idPrint = tnode.id ? `id=${tnode.id}` : null;
  const classesPrint = tnode.classes?.length
    ? `classes={[${tnode.classes.join(', ')}]}`
    : null;
  const dataPrint =
    tnode.type === 'text' ? `data=${JSON.stringify(tnode.data)}` : null;
  const anchorPrint =
    typeof tnode.attributes.href === 'string'
      ? `href=${JSON.stringify(tnode.attributes.href)}`
      : null;
  const srcPrint =
    typeof tnode.attributes.src === 'string'
      ? `src=${JSON.stringify(tnode.attributes.src)}`
      : null;
  const pstyles = withStyles ? printTableStyles(tnode.styles) : null;
  const stylesPrint = pstyles ? `styles={{${pstyles} }}` : null;
  const detailsPrint = [
    tagPrint,
    unregisteredPrint,
    indexPrint,
    idPrint,
    classesPrint,
    dataPrint,
    anchorPrint,
    srcPrint,
    stylesPrint
  ]
    .filter((p) => p !== null)
    .join(' ');
  return `<${tnode.displayName} ${detailsPrint}`;
}

interface TNodePrintState {
  parentLeftPrefix: string;
  isChild: boolean;
  isLast: boolean;
}

export default function serializeTNode(
  tnode: TNode,
  params: Partial<TNodePrintState & TNodePrintOptions> = {}
): string {
  const {
    parentLeftPrefix = '',
    isChild = false,
    isLast = false,
    withStyles = true,
    withNodeIndex = true,
  } = params;
  const prefix = isChild ? '  ' : '';
  const totalPrefixLeft = parentLeftPrefix + prefix;
  const childrenPrint = tnode.children
    .map((c, i) =>
      serializeTNode(c, {
        parentLeftPrefix: parentLeftPrefix + ' '.padStart(prefix.length, ' '),
        isChild: true,
        isLast: i === tnode.children.length - 1,
        withStyles,
        withNodeIndex
      })
    )
    .join('');
  return `${totalPrefixLeft}${tnodePropertiesString(tnode, withStyles)}${
    childrenPrint
      ? `>\n${childrenPrint}\n${totalPrefixLeft}</${tnode.displayName}>`
      : ' />'
  }${isLast ? '' : '\n'}`;
}
