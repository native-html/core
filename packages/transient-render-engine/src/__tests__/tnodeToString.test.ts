import tnodeToString from '../tnodeToString';
import { translateTreeTest } from '../flow/__tests__/utils';
import { TNode } from '../tree/tree-types';

describe('tnodeToString', () => {
  it('should handle deeply nested tags', () => {
    const result = tnodeToString(
      (translateTreeTest(
        '<div><div><span>A<em>a</em></span><span>B</span></div><span>C<strong>D</strong></span></div>'
      ) as unknown) as TNode
    );
    expect(result).toMatchSnapshot();
  });
  it('should print ids, classes and anchors', () => {
    const result = tnodeToString(
      (translateTreeTest(
        '<a href="https://" id="first" class="special"></a>'
      ) as unknown) as TNode
    );
    expect(result).toMatchSnapshot();
  });
  it('should cut data strings over 24 characters with ellipsis', () => {
    const result = tnodeToString(
      (translateTreeTest(
        '<span>This very long string should be truncated and end with an ellipsis.</span>'
      ) as unknown) as TNode
    );
    expect(result).toMatchSnapshot();
  });
  it('should print styles when withStyles is enabled', () => {
    const result = tnodeToString(
      (translateTreeTest(
        '<span style="color: blue;">Blue!</span>'
      ) as unknown) as TNode,
      {
        withStyles: true
      }
    );
    expect(result).toMatchSnapshot();
  });
  it('should not print styles when withStyles is disabled', () => {
    const result = tnodeToString(
      (translateTreeTest(
        '<span style="color: blue;">Blue!</span>'
      ) as unknown) as TNode,
      {
        withStyles: false
      }
    );
    expect(result).toMatchSnapshot();
  });
});
