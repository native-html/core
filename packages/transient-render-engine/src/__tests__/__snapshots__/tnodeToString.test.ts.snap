// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tnodeToString should cut data strings over 24 characters with ellipsis 1`] = `
"•TText(tag=span,data[\\"This very long string sh\\"…])
"
`;

exports[`tnodeToString should handle deeply nested tags 1`] = `
"•TBlock(tag=div)
 ├─•TBlock(tag=div)
 │ ├─•TPhrasing(tag=span)
 │ │ ├─•TText(anonymous,data[\\"A\\"])
 │ │ └─•TText(tag=em,data[\\"a\\"])
 │ └─•TText(tag=span,data[\\"B\\"])
 └─•TPhrasing(tag=span)
   ├─•TText(anonymous,data[\\"C\\"])
   └─•TText(tag=strong,data[\\"D\\"])
"
`;

exports[`tnodeToString should print ids, classes and anchors 1`] = `
"•TText(tag=a,id=first,classes=special,data[\\"\\"],anchor[https://])
"
`;
